src/dwfl_module.cc

```diff
@@ -10,62 +10,75 @@ extern "C" {
 }
 
 namespace ddprof {
-Dwfl_Module *update_module(Dwfl *dwfl, ProcessAddress_t pc, const Dso &dso,
-                           const FileInfoValue &fileInfoValue) {
+DDProfMod update_module(Dwfl *dwfl, ProcessAddress_t pc, const Dso &dso,
+                        const FileInfoValue &fileInfoValue) {
   if (!dwfl)
-    return nullptr;
+    return DDProfMod();
 
   if (fileInfoValue._errored) {
     LG_DBG("DSO Previously errored - mod (%s)", dso._filename.c_str());
-    return nullptr;
+    return DDProfMod();
   }
   // Now that we've confirmed a separate lookup for the DSO based on procfs
   // and/or perf_event_open() "extra" events, we do two things
   // 1. Check that dwfl has a cache for this PID/pc combination
   // 2. Check that the given cache is accurate to the DSO
 
-  Dwfl_Module *mod = dwfl_addrmodule(dwfl, pc);
-  if (mod) {
-    Dwarf_Addr vm_addr;
-    dwfl_module_info(mod, 0, &vm_addr, 0, 0, 0, 0, 0);
-    if (vm_addr != dso._start - dso._pgoff) {
+  DDProfMod ddprof_mod;
+
+  ddprof_mod._mod = dwfl_addrmodule(dwfl, pc);
+  Dwarf_Addr low_addr = 0;
+  Dwarf_Addr high_addr = 0;
+
+  if (ddprof_mod._mod) {
+    const char *main_name = nullptr;
+    dwfl_module_info(ddprof_mod._mod, 0, &ddprof_mod._low_addr,
+                     &ddprof_mod._high_addr, 0, 0, &main_name, 0);
+    if (ddprof_mod._low_addr != dso._start - dso._pgoff) {
       // not using only addresses (as we get off by 1 page errors)
       const std::string &filepath = fileInfoValue.get_path();
-      if (filepath.empty()) {
-        mod = nullptr;
+      if (filepath.empty() || main_name == nullptr) {
+        ddprof_mod._mod = nullptr;
       } else {
+        // Although the check is slightly fragile
+        // we are not comparing the full path as we use the proc maps
+        // to access them. So the root path could change while
+        // this would still be the same file.
         const char *dso_name = strrchr(filepath.c_str(), '/') + 1;
-        if (mod && strcmp(dso_name, mod->name) != 0) {
-          LG_NTC("Incoherent DSO (%s) %lx != %lx dwfl_module (%s)", dso_name,
-                 dso._start - dso._pgoff, vm_addr, mod->name);
-          mod = nullptr;
+        const char *mod_name = strrchr(main_name, '/') + 1;
```

Comment:
Are we assured that `main_name` will always contain a slash?  I'm reasonably confident that `dso_name` will, since both perf_event_open and /proc report absolute paths.

Reply:
Very good question. I'm testing things out right now.

Reply:
I testing things out and could not see it happen. However if we don't guard against it, we will dump as null+1 will be considered as a string by strcmp.

Reply:
Right--doesn't hurt to be cautious if we don't have a strong guarantee from whatever gets the path.  Thanks for looking at that!